x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,421 @@
+        ShardChangesAction.Request request = new ShardChangesAction.Request(params.getLeaderShardId());
+        request.setMinSeqNo(from);
+        request.setMaxSeqNo(to);
+        request.setMaxTranslogsBytes(params.getMaxTranslogBytes());
+        leaderClient.execute(ShardChangesAction.INSTANCE, request, new ActionListener<ShardChangesAction.Response>() {
+            @Override
+            public void onResponse(ShardChangesAction.Response response) {
+                handler.accept(response);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                errorHandler.accept(e);
+            }
+        });
     }
 
     @Override
     protected void onCancelled() {
         markAsCompleted();
     }
 
-    public boolean isRunning() {
-        return isCancelled() == false && isCompleted() == false;
-    }
-
-    void updateProcessedGlobalCheckpoint(long processedGlobalCheckpoint) {
-        this.processedGlobalCheckpoint.set(processedGlobalCheckpoint);
+    protected boolean isStopped() {
+        return isCancelled() || isCompleted();
     }
 
     @Override
-    public Task.Status getStatus() {
-        return new Status(processedGlobalCheckpoint.get());
+    public Status getStatus() {
+        return new Status(processedGlobalCheckpoint, numConcurrentReads, numConcurrentWrites);
     }
 
     public static class Status implements Task.Status {
 
         public static final String NAME = "shard-follow-node-task-status";
 
         static final ParseField PROCESSED_GLOBAL_CHECKPOINT_FIELD = new ParseField("processed_global_checkpoint");
+        static final ParseField NUMBER_OF_CONCURRENT_READS_FIELD = new ParseField("number_of_concurrent_reads");
+        static final ParseField NUMBER_OF_CONCURRENT_WRITES_FIELD = new ParseField("number_of_concurrent_writes");
 
         static final ConstructingObjectParser<Status, Void> PARSER =
-                new ConstructingObjectParser<>(NAME, args -> new Status((Long) args[0]));
+                new ConstructingObjectParser<>(NAME, args -> new Status((long) args[0], (int) args[1], (int) args[2]));
 
         static {
             PARSER.declareLong(ConstructingObjectParser.constructorArg(), PROCESSED_GLOBAL_CHECKPOINT_FIELD);
+            PARSER.declareInt(ConstructingObjectParser.constructorArg(), NUMBER_OF_CONCURRENT_READS_FIELD);
+            PARSER.declareInt(ConstructingObjectParser.constructorArg(), NUMBER_OF_CONCURRENT_WRITES_FIELD);
         }
 
         private final long processedGlobalCheckpoint;
+        private final int numberOfConcurrentReads;
+        private final int numberOfConcurrentWrites;
 
-        Status(long processedGlobalCheckpoint) {
+        Status(long processedGlobalCheckpoint, int numberOfConcurrentReads, int numberOfConcurrentWrites) {
             this.processedGlobalCheckpoint = processedGlobalCheckpoint;
+            this.numberOfConcurrentReads = numberOfConcurrentReads;
```

Comment:
please make sure they are positive or don't use vint to serialize them

Reply:
there is validation for that in the api that creates the persistent task